gorm-tools/src/main/groovy/gorm/tools/repository/BulkableRepo.groovy

```diff
@@ -0,0 +1,162 @@
+
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.core.GenericTypeResolver
+
+import gorm.tools.async.AsyncSupport
+import gorm.tools.job.JobRepoTrait
+import gorm.tools.job.JobTrait
+import gorm.tools.json.Jsonify
+import gorm.tools.support.Results
+
+/**
+ * A trait that allows to insert or update many (bulk) records<D> at once and create Job <J>
+ */
+@CompileStatic
+trait BulkableRepo<D, J extends JobTrait>  {
+
+    @Autowired(required = false)
+    JobRepoTrait jobRepo
+
+    @Autowired
+    @Qualifier("asyncSupport")
+    AsyncSupport asyncSupport
+
+    @Value('${hibernate.jdbc.batch_size:50}')
+    int batchSize  //XXX https://github.com/9ci/domain9/issues/331  test if @Value works on trait
+
+    @Value('${nine.autocash.parallelProcessing.enabled:false}')
+    boolean parallelProcessingEnabled //XXX https://github.com/9ci/domain9/issues/331  we might have to move it, here we can have default
+
+    // GormRepo implements it
+    abstract void bindAndCreate(D entity, Map data, Map args)
+
+    // need cleaner way to do transaction, change it
+    abstract gormStaticApi()
+    abstract  Class<D> getEntityClass()
+
+    //// XXX https://github.com/9ci/domain9/issues/331 Not sure if needed any more
+    //abstract List bulkCreate()
+
+//////// MOVED FROM GORM REPO
+
+    /**
+     * allows to pass in bulk of records at once, for example /api/book/bulk
+     * Each call creates a job that stores info for the call and is returned with results
+     * @param dataList the list of data maps to create
+     * @param args args to pass to doCreate. It can have:
+     *      onError // 'rollback' or 'continue' (catches error) or 'skip'
+     *      errorThreshold // number of errors to stop and rollback
+     *      jdbc.batch_size // all, default or specify number
+     *      gpars.poolsize
+     * @return Job
+     */
+    J bulkCreate(List<Map> dataList, Map args = [:]){
+
+        // create job
+        Class<J> jobClass = (Class<J>)GenericTypeResolver.resolveTypeArguments(getClass(), BulkableRepo)[1]
+        J job = jobClass.newInstance() as J
+
+        // XXX We can do something similar for error count
+        AtomicInteger count = new AtomicInteger(0)
+
+        // for error handling -- based on `onError` from args commit success and report the failure or fail them all
+        //  also use errorThreshold, for example if it failed on more than 10 records we stop and rollback everything
+        //@jdabal - for parallel async, we can not rollback batches which are already commited
+        //as each batch would be in its own transaction.
+        List bulkResult = []
+        if(parallelProcessingEnabled) {
+            //run the batch parallel in batched transactions
+            asyncSupport.parallel([batchSize:batchSize], dataList.collate(batchSize)) { List<Map> batch ->
```

Comment:
should ALWAYS use getters to access props in a trait. getBatchSize() for example. other wise they will always use whats in the trait and not look at what might get overriden